nomad/state/state_store.go

```diff
@@ -467,6 +559,50 @@ func (s *StateStore) JobsByIDPrefix(ws memdb.WatchSet, id string) (memdb.ResultI
 	return iter, nil
 }
 
+// JobVersionsByID returns all the tracked versions of a job.
+func (s *StateStore) JobVersionsByID(ws memdb.WatchSet, id string) ([]*structs.Job, error) {
+	txn := s.db.Txn(false)
+	return s.jobVersionByID(txn, &ws, id)
+}
+
+// jobVersionByID is the underlying implementation for retrieving all tracked
+// versions of a job and is called under an existing transaction. A watch set
+// can optionally be passed in to add the job histories to the watch set.
+func (s *StateStore) jobVersionByID(txn *memdb.Txn, ws *memdb.WatchSet, id string) ([]*structs.Job, error) {
+	// Get all the historic jobs for this ID
+	iter, err := txn.Get("job_versions", "id_prefix", id)
+	if err != nil {
+		return nil, err
+	}
+
+	if ws != nil {
+		ws.Add(iter.WatchCh())
+	}
+
+	var all []*structs.Job
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			break
+		}
+
+		// Ensure the ID is an exact match
+		j := raw.(*structs.Job)
+		if j.ID != id {
+			continue
+		}
+
+		all = append(all, j)
+	}
+
+	// Sort with highest versions first
+	sort.Slice(all, func(i, j int) bool {
```

Comment:
Since we are walking the index, isn't it already sorted? If it's low-to-high, a simple reverse might be sufficient.